client-go/transport: structured, contextual logging

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The revised logging emits one log entry at the start of round-tripping ("Request") and another at the end ("Response"). This avoids the risk that related output gets interleaved by other output.

No API changes are necessary. A contextual logger is picked up from the context of the request that is being handled. The verbosity level of that logger is checked to determine what is supposed to be logged. This enables reducing log details on a by-request basis by storing a `logger.V(1)` in the context of the request.

#### Special notes for your reviewer:

##### request/response logging before/after

###### -v=5 before

```
<no output>
```

###### -v=5 after

```
<no logging>
```

###### -v=6 before

```
I0122 17:00:17.906990  346330 round_trippers.go:560] GET https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset 200 OK in 3 milliseconds
```

###### -v=6 after

```
I0122 17:09:32.381792  360675 round_trippers.go:630] "Response" verb="GET" url="https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset" status="200 OK" milliseconds=3
```

###### -v=7 before

```
I0122 17:00:18.081901  346351 round_trippers.go:470] GET https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset
I0122 17:00:18.081925  346351 round_trippers.go:476] Request Headers:
I0122 17:00:18.081942  346351 round_trippers.go:480]     Accept: application/json
I0122 17:00:18.081953  346351 round_trippers.go:480]     User-Agent: kubectl/v1.33.0 (linux/amd64) kubernetes/bd55d18
I0122 17:00:18.085132  346351 round_trippers.go:581] Response Status: 200 OK in 3 milliseconds
```

###### -v=7 after

```
I0122 17:09:32.570507  360704 round_trippers.go:525] "Request" verb="GET" url="https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset" headers=<
	Accept: application/json
	User-Agent: kubectl/v1.33.0 (linux/amd64) kubernetes/0de6992
 >
I0122 17:09:32.573856  360704 round_trippers.go:630] "Response" status="200 OK" milliseconds=3
```

###### -v=8 before

```
I0122 17:00:18.273097  346374 helper.go:105] "Request Body" body=""
I0122 17:00:18.273166  346374 round_trippers.go:470] GET https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset
I0122 17:00:18.273176  346374 round_trippers.go:476] Request Headers:
I0122 17:00:18.273192  346374 round_trippers.go:480]     Accept: application/json
I0122 17:00:18.273203  346374 round_trippers.go:480]     User-Agent: kubectl/v1.33.0 (linux/amd64) kubernetes/bd55d18
I0122 17:00:18.276270  346374 round_trippers.go:581] Response Status: 200 OK in 3 milliseconds
I0122 17:00:18.276295  346374 round_trippers.go:584] Response Headers:
I0122 17:00:18.276310  346374 round_trippers.go:587]     Content-Length: 2918
I0122 17:00:18.276322  346374 round_trippers.go:587]     Date: Wed, 22 Jan 2025 16:00:18 GMT
I0122 17:00:18.276329  346374 round_trippers.go:587]     Audit-Id: d01b8b74-2c39-4466-bb76-e28bf5cf1a99
I0122 17:00:18.276336  346374 round_trippers.go:587]     Cache-Control: no-cache, private
I0122 17:00:18.276341  346374 round_trippers.go:587]     Content-Type: application/json
I0122 17:00:18.276348  346374 round_trippers.go:587]     X-Kubernetes-Pf-Flowschema-Uid: fbe1061a-6a71-44e6-bf6e-8b7313b60122
I0122 17:00:18.276354  346374 round_trippers.go:587]     X-Kubernetes-Pf-Prioritylevel-Uid: c860c5bd-6097-4d6b-9a18-fb9099a6a480
I0122 17:00:18.276454  346374 helper.go:105] "Response Body" body="{\"kind\":\"DaemonSet\",... [truncated 1894 chars]"
```

###### -v=8 after

```
I0122 17:15:07.759356  368870 helper.go:105] "Request Body" body=""
I0122 17:15:07.759437  368870 round_trippers.go:527] "Request" verb="GET" url="https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset" headers=<
	Accept: application/json
	User-Agent: kubectl/v1.33.0 (linux/amd64) kubernetes/0de6992
 >
I0122 17:15:07.762713  368870 round_trippers.go:632] "Response" status="200 OK" headers=<
	Audit-Id: 85804509-da0e-4f4d-8edd-cf5c01dd9a02
	Cache-Control: no-cache, private
	Content-Length: 2918
	Content-Type: application/json
	Date: Wed, 22 Jan 2025 16:15:07 GMT
	X-Kubernetes-Pf-Flowschema-Uid: fbe1061a-6a71-44e6-bf6e-8b7313b60122
	X-Kubernetes-Pf-Prioritylevel-Uid: c860c5bd-6097-4d6b-9a18-fb9099a6a480
 > milliseconds=3
I0122 17:15:07.762848  368870 helper.go:105] "Response Body" body="{\"kind\":\"DaemonSet\",... [truncated 1894 chars]"
```

###### -v=9 before

```
I0122 17:00:18.460891  346396 helper.go:105] "Request Body" body=""
I0122 17:00:18.460975  346396 round_trippers.go:473] curl -v -XGET  -H "User-Agent: kubectl/v1.33.0 (linux/amd64) kubernetes/bd55d18" -H "Accept: application/json" 'https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset'
I0122 17:00:18.464151  346396 round_trippers.go:560] GET https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset 200 OK in 3 milliseconds
I0122 17:00:18.464178  346396 round_trippers.go:577] HTTP Statistics: GetConnection 0 ms ServerProcessing 2 ms Duration 3 ms
I0122 17:00:18.464193  346396 round_trippers.go:584] Response Headers:
I0122 17:00:18.464208  346396 round_trippers.go:587]     Audit-Id: 25fc0fc2-609b-42d1-8506-b92d5eeceb0a
I0122 17:00:18.464216  346396 round_trippers.go:587]     Cache-Control: no-cache, private
I0122 17:00:18.464222  346396 round_trippers.go:587]     Content-Type: application/json
I0122 17:00:18.464232  346396 round_trippers.go:587]     X-Kubernetes-Pf-Flowschema-Uid: fbe1061a-6a71-44e6-bf6e-8b7313b60122
I0122 17:00:18.464239  346396 round_trippers.go:587]     X-Kubernetes-Pf-Prioritylevel-Uid: c860c5bd-6097-4d6b-9a18-fb9099a6a480
I0122 17:00:18.464244  346396 round_trippers.go:587]     Content-Length: 2918
I0122 17:00:18.464251  346396 round_trippers.go:587]     Date: Wed, 22 Jan 2025 16:00:18 GMT
I0122 17:00:18.464318  346396 helper.go:105] "Response Body" body=<
	{"kind":"DaemonSet",...
 >
```

###### -v=9 after

```
I0122 17:15:07.955594  368893 helper.go:105] "Request Body" body=""
I0122 17:15:07.955683  368893 round_trippers.go:527] "Request" curlCommand=<
	curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.33.0 (linux/amd64) kubernetes/0de6992" 'https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset'
 >
I0122 17:15:07.959011  368893 round_trippers.go:632] "Response" verb="GET" url="https://localhost:6444/apis/apps/v1/namespaces/default/daemonsets/my-daemonset" status="200 OK" headers=<
	Audit-Id: 13081364-925f-4fe6-8e8f-d7b210dbe733
	Cache-Control: no-cache, private
	Content-Length: 2918
	Content-Type: application/json
	Date: Wed, 22 Jan 2025 16:15:07 GMT
	X-Kubernetes-Pf-Flowschema-Uid: fbe1061a-6a71-44e6-bf6e-8b7313b60122
	X-Kubernetes-Pf-Prioritylevel-Uid: c860c5bd-6097-4d6b-9a18-fb9099a6a480
 > milliseconds=3 getConnectionMilliseconds=0 serverProcessingMilliseconds=3
I0122 17:15:07.959095  368893 helper.go:105] "Response Body" body=<
	{"kind":"DaemonSet",...
 >
```


#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
client-go/transport: structured, contextual logging

The revised logging emits one log entry at the start of
round-tripping ("Request") and another at the end ("Response"). This avoids the
risk that related output gets interleaved by other output.

No API changes are necessary. A contextual logger is picked up from the context
of the request that is being handled. The verbosity level of that logger is
checked to determine what is supposed to be logged. This enables reducing log
details on a by-request basis by storing a `logger.V(1)` in the context of the
request.

As before, logging only gets injected into request processing at -v6 or higher,
so normally there is no additional overhead.

pohly commented:
/wg structured-logging

soltysh reviewed (approved):
/lgtm
/approve

/triage accepted
/priority important-longterm

pohly commented:
/assign @aojea 

Picked by the bot.

/assign @sttts 

Has context for this work.

@sttts: should I route all of these PRs through you for approval?

liggitt commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -556,40 +583,67 @@ func (rt *debuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
+		headers := make(headersMap, len(reqInfo.ResponseHeaders))
+		for key, values := range reqInfo.ResponseHeaders {
+			for _, value := range values {
+				headers[key] = value
```
same comment about this stomping multi-value headers with only the last value

liggitt commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -556,40 +583,67 @@ func (rt *debuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
-				reqInfo.DNSLookup.Nanoseconds()/int64(time.Millisecond),
-				reqInfo.Dialing.Nanoseconds()/int64(time.Millisecond),
-				reqInfo.TLSHandshake.Nanoseconds()/int64(time.Millisecond),
+			kvs = append(kvs,
```
nit: I don't know the expectations for structured key casing, but maybe `dnsLookupMilliseconds` and `tlsHandshakeMilliseconds`?

pohly replied:
Usually it's "start lowercase", except for acronyms. It's just a guideline and what you suggested is more readable - let's change it.

liggitt commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -556,40 +583,67 @@ func (rt *debuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
+	kvs = make([]any, 0, 10)
+	if levels&(1<<DebugURLTiming) != 0 {
+		kvs = append(kvs, "verb", reqInfo.RequestVerb, "url", reqInfo.RequestURL)
+	}
```
did we lose reqInfo.RequestURL? I don't see that included in kvs

pohly replied:
Huh? It's right here in line 588.

Or did you mean status? That gets added below in line 591. Duration is in 603.

The if checks should be so that the same information is included for each constant.

liggitt replied:
> Huh? It's right here in line 588.

🤦  apparently I'm unable to read more than two args to a function

liggitt commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -556,40 +583,67 @@ func (rt *debuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
-		stats += fmt.Sprintf(` Duration %d ms`, reqInfo.Duration.Nanoseconds()/int64(time.Millisecond))
-		klog.Infof("HTTP Statistics: %s", stats)
 	}
+	logger.Info("Response", kvs...)
```
do you have a before/after of v6/v7/v8/v9 log output in normal text logs using the old and new construction?

pohly replied:
Will add that.

liggitt replied:
thanks, that's helpful. I updated the description to put the before/after for each level next to each other for easier comparison

liggitt commented on staging/src/k8s.io/client-go/transport/token_source.go:
```diff
@@ -182,6 +182,8 @@ func (ts *cachingTokenSource) Token() (*oauth2.Token, error) {
 		if ts.tok == nil {
 			return nil, err
 		}
+		// TODO: avoid logging here - but not sure how.
```
connect TODOs to an issue we can go to for context / assignment / tracking

pohly replied:
Or remove this TODO and decide now that we just leave this call as it is?

If I were to create an issue, it would probably never get addressed and just cause constant work for re-triaging until eventually it gets auto-closed.

liggitt replied:
fine with me

pohly replied:
After thinking about this again, I used `klog.TODO().Error(err, "Unable to rotate token")` with a long comment explaining why there is no tracking issue.

I could have used `klog.Background()`, but that isn't quite right: if someone finds a way to pass in a logger, that line should be changed. I've encountered quite a few places where developers used `context.Background()` in places where passing in a context would have been the right thing to do; this here would be similar.

pohly commented on staging/src/k8s.io/client-go/transport/cache.go:
```diff
@@ -116,10 +117,13 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
 		tlsConfig.GetClientCertificate = dynamicCertDialer.GetClientCertificate
 		dial = dynamicCertDialer.connDialer.DialContext
-		go dynamicCertDialer.Run(DialerStopCh)
+		go dynamicCertDialer.run(DialerStopCh)
```
When I saw the `Run` method my first impression was "this must be the implementation of some interface". It turned out that it isn't. With `run` I am certain that this is completely internal to the package.

Not important. I can change it back.

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -68,19 +70,14 @@ func HTTPWrappersForConfig(config *Config, rt http.RoundTripper) (http.RoundTrip
-	case bool(klog.V(6).Enabled()):
-		rt = NewDebuggingRoundTripper(rt, DebugURLTiming)
+	//nolint:logcheck // The actual logging is done with a different logger, so only checking here is okay.
+	if klog.V(6).Enabled() {
```
Let me try that. I wasn't sure whether some additional overhead would be acceptable.

pohly replied:
Changed. I am keeping it in a separate commit, in case that we change out mind about it.

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -464,27 +462,56 @@ func maskValue(key string, value string) string {
 }
 
 func (rt *debuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	logger := klog.FromContext(req.Context())
```
Yes.

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -464,27 +462,56 @@ func maskValue(key string, value string) string {
 			for _, value := range values {
 				value = maskValue(key, value)
-				klog.Infof("    %s: %s", key, value)
+				headers[key] = value
```
True. I'll change to a list of pairs.

pohly replied:
It occurred to me that I can use the JSON logger built into logr to test that marshaling as JSON works as intended, without creating a dependency on zap. I've updated the unit test accordingly and also added a test case for a multi-valued header.

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -464,27 +462,56 @@ func maskValue(key string, value string) string {
 
-	if rt.levels[DebugJustURL] {
-		klog.Infof("%s %s", reqInfo.RequestVerb, reqInfo.RequestURL)
+	kvs := make([]any, 0, 8)
```
Because I wanted to leave some room in case someone adds something else and forgets to bump this up. It doesn't make a difference because this gets backed by the stack.

I'll switch to an exact count with a comment that reminds future developers to increase it.

pohly replied:
Scratch that. After taking a close look: the first append adds "verb" and "url", so 8 is exact after all. Reformatted with multiple lines to make it more obvious and added a comment.

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -464,27 +462,56 @@ func maskValue(key string, value string) string {
 		}
+		kvs = append(kvs, "headers", headers)
 	}
+	logger.Info("Request", kvs...)
```
I can add that to the PR description.

pohly replied:
Done. A useful exercise as it gave me ideas for making the output more useful :sweat_smile: 

- Don't log records with just a Request or Response message text. 
- Log the curl command as multi-line string for easier cut-and-paste and to make it more readable (no quoting).

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -412,17 +409,18 @@ const (
 	}
 	for _, v := range levels {
-		drt.levels[v] = true
+		drt.levels |= 1 << v
```
I hadn't considered that I am not introducing these constants in this PR.

Changing these constants is an API break. I prefer not to do that when the only motivation is cleanup. It's okay if someone just uses them as parameters to `NewDebuggingRoundTripper`. But perhaps someone else also uses them for a bit field and would get broken by increasing the values.

pohly replied:
It's not that bad. This doesn't compile because `int` (bit field) != `DebugLevel` (constant):
```
var x int
if x & DebugLevel { ... }
```

pohly commented on staging/src/k8s.io/client-go/transport/round_trippers.go:
```diff
@@ -464,27 +462,56 @@ func maskValue(key string, value string) string {
+		if loggerV := logger.V(9); loggerV.Enabled() {
+			logger = loggerV
+			// The curl command replaces logging of the URL.
+			levels |= (1 << DebugCurlCommand) | (1 << DebugURLTiming) | (1 << DebugDetailedTiming) | (1 << DebugResponseHeaders)
```
Let's combine. I cannot quite nail where someone might want to do `NewDebuggingRoundTripper(DebugByContext, DebugJustURL)`, but it's not inconceivable. It's basically a "always log the URL, then add more information depending on the context verbosity". I added more explicit documentation and a test case.

pohly commented:
I pushed the update: https://github.com/kubernetes/kubernetes/compare/2a165a8f396a066ab7a0743947c93277784920bd..cb741b2d322609f25e61f55409d7eec92e0ff720

liggitt commented:
unit test failure looks like a fragile test that needs adjusting:
```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
index 992eef77a18..e9b33a3473c 100644
--- a/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator_test.go
@@ -948,7 +948,7 @@ func v4WriteStatusFunc(stream io.Writer) func(status *apierrors.StatusError) err
 	}
 }
 
-func fakeTransport() (*http.Transport, error) {
+func fakeTransport() (http.RoundTripper, error) {
 	cfg := &transport.Config{
 		TLS: transport.TLSConfig{
 			Insecure: true,
@@ -959,9 +959,5 @@ func fakeTransport() (*http.Transport, error) {
 	if err != nil {
 		return nil, err
 	}
-	t, ok := rt.(*http.Transport)
-	if !ok {
-		return nil, fmt.Errorf("unknown transport type: %T", rt)
-	}
-	return t, nil
+	return rt, nil
 }
```

liggitt commented:
I went over the implementation changes pretty carefully, and the before / after in the description was helpful. I still think the in-place bitshifting makes stuff harder to read than it needs to be, and I think we could switch the declared constants to be bit-shifted without compatibility issues, but I don't care *that* much.

I'd like a second ack (maybe from @deads2k or @sttts?) on unconditionally wrapping the roundtripper in the debug roundtripper, and moving the v(6) check to be inside the debug round tripper... I guess the unit test failure in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129330/pull-kubernetes-unit/1882123626438725632 is an example of observable behavior difference caused by this that could disrupt existing (arguably fragile buggy) code in the real world.

pohly commented:
> I'd like a second ack (maybe from @deads2k or @sttts?) on unconditionally wrapping the roundtripper in the debug roundtripper, and moving the v(6) check to be inside the debug round tripper...

For reference, that is currently commit 4901d8b. I need to update it to fix the unit test failure in `k8s.io/apiserver/pkg/util`.

> I guess the unit test failure in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129330/pull-kubernetes-unit/1882123626438725632 is an example of observable behavior difference caused by this that could disrupt existing (arguably fragile buggy) code in the real world.

True. That it isn't limited to unit tests in the `transport` package shows that it is an external change, although again limited to unit tests and thus perhaps something that no sane code will do.

pohly commented:
This error is also caused by wrapping:
```
=== RUN   TestRestClientTimeout
2025/01/22 18:28:37 http: response.Write on hijacked connection from fmt.Fprintf (print.go:225)
    connection_test.go:368: timeout error expected, received Get "http://127.0.0.1:45231/?timeout=1s": net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

It's a flake which I can reproduce with this PR, but not with master and not without wrapping.

I think it's a race between `CancelRequest` (causing the unexpected "request canceled") and a context timeout in the caller. 

The expected error is:
```
error(*net/url.Error) *{
	Op: "Get",
	URL: "http://127.0.0.1:34051/?timeout=1s",
	Err: error(context.deadlineExceededError) {},}
```

The unexpected one is:
```
error(*net/url.Error) *{
	Op: "Get",
	URL: "http://127.0.0.1:36249/?timeout=1s",
	Err: error(*net/http.timeoutError) *{
		err: "net/http: request canceled (Client.Timeout exceeded while awaiting headers)",},}
```

At first glance it looks like the test should accept both, because both look legitimate to me. But this did not happen before. That's because `http.Client` checks for roundtrippers which are [known](https://cs.opensource.google/go/go/+/master:src/net/http/client.go;l=315-329) and with them [uses](https://cs.opensource.google/go/go/+/master:src/net/http/client.go;l=358-367) only context cancellation.

Our wrapper is not known and provides `CancelRequest`, so that now gets called and thus also the underlying `http.Transport.CancelRequest`.

I tried to make our code smarter: if and only if the wrapped roundtripper has `CancelRequest` and is not one of those known to need it, then provide `CancelRequest` in the wrapper. This could almost have worked, except for https://github.com/kubernetes/kubernetes/blob/06331ac573a182bdace8f238ca94a1f476958b21/staging/src/k8s.io/client-go/transport/token_source.go#L117-L123 which has additional logic where it calls `CancelRequest`.

Perhaps we can finally assume that all roundtrippers support context cancellation and remove CancelRequest from all of our wrappers? That should also get rid of the `deprecated: golang.org/x/oauth2: Transport.CancelRequest no longer does anything; use contexts` log message during unit testing that I [mentioned earlier](https://github.com/kubernetes/kubernetes/pull/129330#issuecomment-2607146932) (occurs with and without this PR).

But not in this PR... Here I am going to accept both errors to avoid the flake: https://github.com/kubernetes/kubernetes/compare/4901d8bc19392ef92a05e126a51f82b9dcdec987..f52481ff58b49933c3947a24f87f1d77816ab4da

pohly commented:
/test pull-kubernetes-e2e-gce 

Failed to start, no logs?

/test pull-kubernetes-node-e2e-containerd

Image pull error.

pohly commented:
As seen in our own unit tests and in [the standard library](https://github.com/kubernetes/kubernetes/pull/129330#issuecomment-2609977702), wrapping the underlying transport makes a difference for consumers. Therefore I am leaning towards the [original behavior](https://github.com/kubernetes/kubernetes/pull/129330#discussion_r1923986652) of only wrapping when the global log verbosity is >= 6.

To move forward with this PR, I've:
- Added a comment that the current behavior might change ("currently wraps only at >= 6, might change") and removed the unconditional wrapping: https://github.com/kubernetes/kubernetes/compare/f52481ff58b49933c3947a24f87f1d77816ab4da..a85f489b28d3b0ef82dffb267b6145c73c2d0e33
- Moved the commit which makes wrapping unconditional to a separate PR: https://github.com/kubernetes/kubernetes/pull/129831.

We can continue to discuss the merit of that particular change there.

> Tag them in for a second set of eyes, give until Monday to raise any issues, then merge if none are raised?

No additional opinions so far, shall we move ahead and merge, @liggitt ?

liggitt commented:
/lgtm
/approve
